### ISSUE TITLE
`model.forward` requires `num_logits_to_keep`, not `logits_to_keep`

### ISSUE THREAD

**Issue by richardwth** [huggingface/trl#2770]
### Reproduction

In `_get_per_token_logps` method of `grpo_trainer.py`, the model is called as
```python
logits = model(
    input_ids=input_ids, attention_mask=attention_mask, logits_to_keep=logits_to_keep + 1
).logits  # (B, L, V)
```
But on Transformers 4.48, Qwen 2's `model.forward` has no `logits_to_keep` argument. It should be `num_logits_to_keep`, i.e.,
```python
logits = model(
    input_ids=input_ids, attention_mask=attention_mask, num_logits_to_keep=num_logits_to_keep + 1
).logits  # (B, L, V)
```

### System Info

- Platform: Linux-5.10.134-16.3.al8.x86_64-x86_64-with-glibc2.35
- Python version: 3.10.16
- PyTorch version: 2.5.1
- CUDA device(s): NVIDIA L20Z, NVIDIA L20Z, NVIDIA L20Z, NVIDIA L20Z, NVIDIA L20Z, NVIDIA L20Z, NVIDIA L20Z, NVIDIA L20Z
- Transformers version: 4.48.0
- Accelerate version: 1.2.1
- Accelerate config: not found
- Datasets version: 3.2.0
- HF Hub version: 0.27.1
- TRL version: 0.15.0.dev0
- bitsandbytes version: 0.45.0
- DeepSpeed version: 0.16.2
- Diffusers version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: 1.60.0
- PEFT version: 0.6.2

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by qgallouedec** [comment]
What model do you use?

**Comment by richardwth** [comment]
> What model do you use?

I am using Qwen2.5 7B and Llama3 8B.

**Comment by qgallouedec** [comment]
Can't reproduce 🤔 (with both transformers 4.49 dev and 4.48):

```python
>>> from transformers import AutoModelForCausalLM
>>> import torch
>>> model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-7B").to("cuda")
>>> input_ids = torch.randint(100, 200, (4, 256), device="cuda")
>>> model(input_ids, logits_to_keep=128)
CausalLMOutputWithPast(loss=None, logits=tensor([[[ 3.1937,  5.0300,  5.1271,  ...,  0.9952,  0.9950,  0.9953],
...
         [-5.3295, -7.6368, -1.7083,  ...,  4.3898,  4.3899,  4.3898]]],
       device='cuda:0', grad_fn=<UnsafeViewBackward0>), past_key_values=DynamicCache(), hidden_states=None, attentions=None)
>>>
```

**Comment by Co1lin** [comment]
@qgallouedec Hey, could you check the shape of the returned logits?

```python
out = model(input_ids, logits_to_keep=128)
out.logits.shape
```

On my side, using `logits_to_keep=128` will return `torch.Size([4, 256, 152064])`. Instead, using `num_logits_to_keep=128` gives `torch.Size([4, 128, 152064])`.

**Comment by Co1lin** [comment]
@qgallouedec Also, do you think this is related to https://github.com/huggingface/trl/issues/2731?

**Comment by qgallouedec** [comment]
Nice finding. Indeed, with 4.48, you'll get 256, while with 4.49, you'll get 128, as expected.
You'll need to clear the cache as well after upgrading or load with `force_download=True`

**Comment by richardwth** [comment]
Right... Transformers' [latest modeling_qwen2.py](https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen2/modeling_qwen2.py) says:
```
@deprecate_kwarg("num_logits_to_keep", version="4.50", new_name="logits_to_keep")
```
So this is a bug related to some models of some Transformers versions. I've modified my issue to reflect this.

**Comment by nopepper** [comment]
!!! Thank you so much

I was baffled and debugging my code for hours because of this. After applying the fix, the difference is obvious:

Before:

![Image](https://github.com/user-attachments/assets/0920c001-a44b-4139-8f0d-b45521daf8db)

After:

![Image](https://github.com/user-attachments/assets/aa228e9c-28de-4160-83e3-865bed5d25e8)

This is a pretty major bug that makes the method unusable for Qwen models (and maybe others) using the latest `transformers` version.

**Comment by Co1lin** [comment]
@nopepper Hey, could you share the versions of `transformers` and `trl` that you are using now? Also, which model are you using?

**Comment by Superskyyy** [comment]
Interestingly even with this bug I was able to train a Qwen model properly...
